Examples/Algorithms/TruthTracking/ActsExamples/TruthTracking/ParticleSmearing.cpp

```diff
@@ -40,64 +41,64 @@ ActsExamples::ProcessCode ActsExamples::ParticleSmearing::execute(
-    const auto pt = particle.transverseMomentum();
-    const auto p = particle.absMomentum();
+  for (auto&& [vtxId, vtxParticles] : groupBySecondaryVertex(particles)) {
+    // a group contains at least one particle by construction. assume that all
+    // particles within the group originate from the same position and use it to
+    // as the refernce position for the perigee frame.
+    auto perigee = Acts::Surface::makeShared<Acts::PerigeeSurface>(
+        vtxParticles.begin()->position());
 
-    // compute momentum-dependent resolutions
-    const auto sigmaD0 =
-        m_cfg.sigmaD0 +
-        m_cfg.sigmaD0PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaD0PtB) * pt);
-    const auto sigmaZ0 =
-        m_cfg.sigmaZ0 +
-        m_cfg.sigmaZ0PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaZ0PtB) * pt);
-    const auto sigmaP = m_cfg.sigmaPRel * p;
-    // var(q/p) = (d(1/p)/dp)² * var(p) = (-1/p²)² * var(p)
-    const auto sigmaQOverP = sigmaP / (p * p);
-    // shortcuts for other resolutions
-    const auto sigmaT0 = m_cfg.sigmaT0;
-    const auto sigmaPhi = m_cfg.sigmaPhi;
-    const auto sigmaTheta = m_cfg.sigmaTheta;
-    // converstion from perigee d0,z0 to curvilinear u,v
-    // d0 and u differ only by a sign
-    const auto sigmaU = sigmaD0;
-    // project from z0 to the second axes orthogonal to the track direction
-    const auto sigmaV = sigmaZ0 * std::sin(theta);
+    for (const auto& particle : vtxParticles) {
+      const auto time = particle.time();
+      const auto phi = Acts::VectorHelpers::phi(particle.unitDirection());
+      const auto theta = Acts::VectorHelpers::theta(particle.unitDirection());
+      const auto pt = particle.transverseMomentum();
+      const auto p = particle.absMomentum();
+      const auto q = particle.charge();
 
-    // draw random noise
-    const auto deltaD0 = sigmaD0 * stdNormal(rng);
-    const auto deltaZ0 = sigmaZ0 * stdNormal(rng);
-    const auto deltaT0 = sigmaT0 * stdNormal(rng);
-    const auto deltaPhi = sigmaPhi * stdNormal(rng);
-    const auto deltaTheta = sigmaTheta * stdNormal(rng);
-    const auto deltaP = sigmaP * stdNormal(rng);
+      // compute momentum-dependent resolutions
+      const double sigmaD0 =
+          m_cfg.sigmaD0 +
+          m_cfg.sigmaD0PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaD0PtB) * pt);
+      const double sigmaZ0 =
+          m_cfg.sigmaZ0 +
+          m_cfg.sigmaZ0PtA * std::exp(-1.0 * std::abs(m_cfg.sigmaZ0PtB) * pt);
+      const double sigmaP = m_cfg.sigmaPRel * p;
+      // var(q/p) = (d(1/p)/dp)² * var(p) = (-1/p²)² * var(p)
+      const double sigmaQOverP = sigmaP / (p * p);
+      // shortcuts for other resolutions
+      const double sigmaT0 = m_cfg.sigmaT0;
+      const double sigmaPhi = m_cfg.sigmaPhi;
+      const double sigmaTheta = m_cfg.sigmaTheta;
 
-    // smear the position/time
-    Acts::Vector4D pos4 = particle.position4();
-    pos4[Acts::ePos0] += deltaD0 * std::sin(phi);
```

Comment:
Just to confirm if I understand the difference between the old and new smearing correctly. In the old approach here, if we use the `newPhi` to calculate the `pos4[Acts::ePos0]` and `pos4[Acts::ePos1]`, the smearing should be the same as the new approach (though the parameters are represented with different reference surface), right?

Reply:
The effective smearing should be the same; it is the representation that changes.

The old version would smear in the curvilinear frame of the true particle by transforming the perigee uncertainties into the curvilinear frame. The new version smears directly in the perigee frame at the true particle position; it does not require the transformation and requires us to create fewer reference surfaces (one per particle vs. one per vertex).

Reply:
I see.Yeah, in the new approach, less surfaces are created. Thanks!

Reply:
@XiaocongAi Since you already looked at it, would you mind reviewing this?

Reply:
Sure, I will have a look.